rfcs/0003-gene-sets/api.md

```diff
@@ -0,0 +1,400 @@
+stored in the database for that gene set.
+
+- If the values match, the server will update the gene set as specified in the PUT
+  request and update the counter by incrementing it by 1. The response to the client will include the updated counter value
+  to be used in future PUT requests for that gene set.
+- If the counter stored in the database does not match the counter sent by the client the server will immediately send
+  an error message to the client, specifying the name and uuid of the gene set that is out of sync. The client will
+  then send over the entire gene set including all linked genes as a POST request. Because the gene set name is already
+  taken for that user/dataset, the server will delete the currently stored gene set and any associated genes and recreate
+  it based on the information included in the POST request. This will generate a new UUID for the gene set, and reset the
+  consensus counter.
+  In the case that the server times out or does not respond to a PUT request from the client, the client can retry or
+  just send a POST request to recreate the gene set.
+  Because recreating the gene set assigns it a new UUID, any pending or high latency PUT requests sent prior to the update
+  be ignored because they reference a UUID that no longer exists.
+
+In the desktop version of the application the gene sets (and genes) will be stored locally as a csv under the name of
+the dataset and the user session. On updates the application will create and store a new csv, maintaining the last
+10 updates (this will match the current functionality for annotations in the desktop environment)
+
+### Data Lifecycle
+
+#### Creation
+
+Gene sets are created via a POST request to `/dataset_name/api/v0.1/gene_sets/`. This request is sent
+
+- after a user uploads a csv continaing one (or more) gene sets.
+- based on the results of comparing differential expression of two groups of cells.
+- after the user creates a new gene set via the application GUI
+  Genes are created as a part of the gene set creation step. They can also be added to an existing gene set via a PUT
+  requst to `/dataset_name/api/v0.1/gene_sets/{gene_set_uuid}`
+
+#### Updates
+
+Gene sets are updated via a PUT request to `/dataset_name/api/v0.1/gene_sets/{gene_set_uuid}`. Genes can also be
+updated via this mechanism
+
+#### Deletion
+
+Gene sets are deleted via a DELETE request to `/dataset_name/api/v0.1/gene_sets/{gene_set_uuid}`. All genes attached to
+the gene set will be deleted as well.
+Genes can be removed from a gene set by including their name in the `genes_to_delete` list that is part of the
+request body for PUT requests to `/dataset_name/api/v0.1/gene_sets/{gene_set_uuid}`
+
+### APIs
+
+All APIs assume user identification is available with the request and the dataset being worked on is specified in the url
+Future API endpoints for consideration
```

Comment:
Did you mean for these to be marked as future endpoints?   I think the one to get a list al all gene sets would be necessary.

Reply:
I didnt think they were necessary for the mvp, do you think they are?

Reply:
Aren't these the endpoints that the client uses to get the current gene sets?

Reply:
no. The client can use GET dataset_name/api/v0.1/gene_sets/
That will return all of the genesets associated with the user/dataset